File: lib/src/communication/PluginInformationFactory.dart
import 'dart:collection';

import 'CommunicationSecret.dart';
import 'PluginInformation.dart';

/// <p>A factory class to create plugin information entries.</p>
class PluginInformationFactory {
  static final Map<String, PluginInformation> store = HashMap();

  /// <p>Retrieves plugin information for a plugin.</p>
  /// @param id the id of the the plugin
  /// @param executor the executor string required to run the plugin
  /// @param port     the current port of the plugin (-1 denotes unknown)
  /// @param secret   the communication secret
  /// @return returns the information object or null if not available
  static PluginInformation? getPluginInformation(String id,
      [String? executor, int? port, CommunicationSecret? secret]) {
    if (executor == null || port == null || secret == null) {
      return store[id.toLowerCase()];
    }

    var info = store[id.toLowerCase()];
    if (info == null) {
      info = PluginInformation(executor, port, secret);
      setPluginInformation(id, info);
    }
    return info;
  }

  /// <p>Puts a pluginInformation into the store.</p>
  /// @param id   the id of the the plugin
  /// @param info the information object containing all relevant information to contact the plugin
  /// @return the previously set information or null if new
  static PluginInformation? setPluginInformation(
      String id, PluginInformation info) {
    var old = getPluginInformation(id);
    store[id.toLowerCase()] = info;
    return old;
  }

  /// <p>Clears all plugin information from the store.</p>
  static void zap() {
    store.clear();
  }
}

File: java/GeigerApiPluginConnector/src/main/java/eu/cybergeiger/api/plugin/PluginInformationFactory.java
package eu.cybergeiger.api.plugin;

import java.util.HashMap;
import java.util.Map;

/**
 * <p>A factory class to create plugin information entries.</p>
 */
public class PluginInformationFactory {

  private static final Map<String, PluginInformation> store = new HashMap<>();

  /**
   * <p>Retrieves plugin information for a plugin.</p>
   *
   * @param id the id of the the plugin
   * @param executor the executor string required to run the plugin
   * @param port     the current port of the plugin (-1 denotes unknown)
   * @param secret   the communication secret
   * @return returns the information object or null if not available
   */
  public static PluginInformation getPluginInformation(String id, String executor, int port,
                                                       CommunicationSecret secret) {
    PluginInformation info = store.get(id.toLowerCase());
    if (info == null) {
      info = new PluginInformation(executor, port, secret);
      setPluginInformation(id, info);
    }
    return info;
  }

  /**
   * <p>Retrieves plugin information for a plugin.</p>
   *
   * @param id the id of the the plugin
   * @return returns the information object or null if not available
   */
  public static PluginInformation getPluginInformation(String id) {
    return store.get(id.toLowerCase());
  }

  /**
   * <p>Puts a pluginInformation into the store.</p>
   *
   * @param id   the id of the the plugin
   * @param info the information object containing all relevant information to contact the plugin
   * @return the previously set information or null if new
   */
  public static PluginInformation setPluginInformation(String id, PluginInformation info) {
    PluginInformation old = getPluginInformation(id);
    store.put(id.toLowerCase(), info);
    return old;
  }

  /**
   * <p>Clears all plugin information from the store.</p>
   */
  public static void zap() {
    store.clear();
  }
}
